algorithm.go

```diff
@@ -164,3 +173,23 @@ func (a Algorithm) FromBytes(p []byte) Digest {
 func (a Algorithm) FromString(s string) Digest {
 	return a.FromBytes([]byte(s))
 }
+
+// Validate validates the encoded portion string
+func (a Algorithm) Validate(encoded string) error {
+	r, ok := anchoredEncodedRegexps[a]
+	if !ok {
+		return ErrDigestUnsupported
```

Comment:
This is safer than before, but we can [still][1] do better (assuming go-digest is ok with [the spec restrictions on the encoded charset][2]).  @stevvooe felt that had already been handled in “[the top-level validate][3]”, but `Algorithm.Validate` is a public method, so you can hit it without going through `Digest.Validate`.  I expect we want to handle the following cases:

* Algorithm is in `anchoredEncodedRegexps`, which you already handle correctly.
* Algorithm is not in `anchoredEncodedRegexps` and:
    * The encoded part validates vs. the [generic spec charset][2], in which case you should return `ErrDigestUnsupported` like you're doing here.
    * The encoded part does not validate vs. the generic spec charset, in which case we can return `ErrDigestInvalidFormat` even without recognizing the algorithm.

I'm not clear which case zero-length encoded parts fall into; they should be either `ErrDigestInvalidFormat` (which you've chosen below) or `ErrDigestInvalidLength`.  Either way, the comment for this function should document which applies to that case.

[1]: https://github.com/opencontainers/go-digest/pull/34#discussion_r116304776
[2]: https://github.com/opencontainers/image-spec/blob/f03dbe35d449c54915d235f1a3cf8f585a24babe/descriptor.md#digests
[3]: https://github.com/opencontainers/go-digest/pull/34#discussion_r116338977